# DECLARATION
<br>
(1) I hereby agree to follow any and all assignment rules and procedures as stated in Canvas for this course, MATH2319.

(2) In particular, I solemnly swear that I will not discuss/ have not discussed my assignment solutions with anyone in any way and the solutions I am submitting are my own personal work.

Full Name: [Pranamya Korde]

# Question 1 : Data Preperation for Machine Learning 

## Data Description :

This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset contains 690 number of observation instances with 16 attribute which contain a class atrribute. This dataset contains number of nominal categorical variables with both binary and multivalue variables and Numerical feature variables.

The dataset contains almost 7-8% of missing values spreading across all the columns.Observations may have zero or many missing values.

The class distribution of the data is :

+: 307 (44.5%)
<br>
-: 383 (55.5%)

# Loading DataSet and Summary Statistics

we have downloaded the data file (crx.data) from https://archive.ics.uci.edu/ml/datasets/Credit+Approval and then uploaded it into Azure project to read it as a local file. we have used read_csv function from pandas library to read and load the data. 

As there are no column names mentioned for the data in the dataset to protect the confidentiality, we provide column names as A1 to A16 for all the columns while reading the data.

In [1]:
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

attrnames = ['A1','A2','A3','A4','A5','A6','A7','A8','A9','A10','A11','A12','A13','A14','A15','A16']
df = pd.read_csv('./crx.data', sep = ',', header = None, names = attrnames)
df.head(5)

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

We can check further details about the data as below : 

In [2]:
df.shape

(690, 16)

The above data consists of 690 rows with 15 descriptive features and 'A16' as a target feature(Class attribute)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
A1     690 non-null object
A2     690 non-null object
A3     690 non-null float64
A4     690 non-null object
A5     690 non-null object
A6     690 non-null object
A7     690 non-null object
A8     690 non-null float64
A9     690 non-null object
A10    690 non-null object
A11    690 non-null int64
A12    690 non-null object
A13    690 non-null object
A14    690 non-null object
A15    690 non-null int64
A16    690 non-null object
dtypes: float64(2), int64(2), object(12)
memory usage: 86.3+ KB


The above information shows that, the data contains 4 numerical features and 12 Categorical feature along with the target feature.

To describe numerical and Categorical features differently we use describe method respectively as follows:

In [4]:
df.describe(include = np.number).round(3)

A3       A8      A11         A15
count  690.000  690.000  690.000     690.000
mean     4.759    2.223    2.400    1017.386
std      4.978    3.347    4.863    5210.103
min      0.000    0.000    0.000       0.000
25%      1.000    0.165    0.000       0.000
50%      2.750    1.000    0.000       5.000
75%      7.208    2.625    3.000     395.500
max     28.000   28.500   67.000  100000.000

In [5]:
df.describe(include = np.object).round(3)

A1   A2   A4   A5   A6   A7   A9  A10  A12  A13    A14  A16
count   690  690  690  690  690  690  690  690  690  690    690  690
unique    3  350    4    4   15   10    2    2    2    3    171    2
top       b    ?    u    g    c    v    t    f    f    g  00000    -
freq    468   12  519  519  137  399  361  395  374  625    132  383

# Checking for Missing Values

As suggested in the data discription file crx.names, the dataset contain almost 7% of the missing values. We can check missing values by isna() function in the pandas library.

In [6]:
df.isna().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

As we can see above, all the column contains 0 NaN values but sometimes missing values are replcaed with unusual values in the data like '?' or '9999' etc. So we have to look for those unusual values and set them to the missing values first. To do this, we check the unusual values in the categorical columns first and set them to the missing values.

In [7]:
categorical_cols = df.columns[df.dtypes==object].tolist()
for col in categorical_cols :
    print(df.loc[df[col]=='?',[col]].count())

A1    12
dtype: int64
A2    12
dtype: int64
A4    6
dtype: int64
A5    6
dtype: int64
A6    9
dtype: int64
A7    9
dtype: int64
A9    0
dtype: int64
A10    0
dtype: int64
A12    0
dtype: int64
A13    0
dtype: int64
A14    13
dtype: int64
A16    0
dtype: int64


As we can see above, categorical features A1, A2, A4, A5, A6, A7 and A14 have missing values which are presented as '?'.  We can set all the unsual values which are observed as '?' with 'NaN' values mentioned in the numpy library. we can use simple replace function for it. we can check the null values again after setting unusual values with null(i.e. NaN).

In [8]:
df = df.replace('?', np.NaN)
df.isna().sum()

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

We can see above that, now there are missing values present in those feature columns. we can check it with the same mentioned in the description file (crx.names) of the data file.
<br>
As mentioned in the description of the data preperation we have to impute those missing values with the mode of the respective feature columns. We are using fillna() function mentioned in the pandas library for the imputation. After imputation we can see that the NaN values are imputed with the Mode values of their repspective columns and there are no missing values present in the dataset now.

In [9]:
for col in categorical_cols:
    df[col].fillna(df[col].mode()[0], inplace = True)
df.isna().sum()

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

# Encoding Categorical Features

Prior to modelling, it is necessary to encode all the categorical features (Target as well as Descriptive) into a set of numerical features. We have to encode the target features explicitely by dropping it from rest of the dataset.

## Encoding the Target Feature

We remove target feature 'A16' from the dataframe df and rename it as 'Target'. we store rest of the data in the same data frame df.

In [10]:
Target = df['A16']
df = df.drop(columns=['A16'])
Target.value_counts()

-    383
+    307
Name: A16, dtype: int64

While encoding this feature, we have to assign positive class with value 1 and negative with 0. we will use replace function for that purpose.

In [11]:
Target = Target.replace({'-': 0, '+': 1})
Target.value_counts()

0    383
1    307
Name: A16, dtype: int64

## Encoding Categorical Descriptive Features

We can see from the data that all the descriptive features appears to be nominal as there is no natural ordering present between the values of these features, hence, we have to perform one-hot-encoding for them.
<br>
As mentioned explicitely in the problem, we have to descritize column 'A2' via equal frequency binning with 3 bins named "low", "medium", "high". and then we have to use integer encoding for this feature. we have to copy our dataframe into another dataframe to prevent the data loss in case.

In [12]:
df_cat = df.copy()
df_cat['A2'] = df_cat['A2'].astype(np.float64) 
df_cat['A2'] = pd.qcut(df_cat['A2'], q=3,labels=['low', 'medium', 'high'])
df_cat.head(5)

A1      A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15
0  b  medium  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0
1  a    high  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560
2  a  medium  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824
3  b  medium  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3
4  b     low  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0

We can see above that the values in the numerical descriptive feature 'A2' are populated in the three equal frequency bins namely "low", "medium" and "high". 
<br>
The next step is to encode them by using integer encoding technique. For this we need define a mapping between the levels and the integers using a dictionary.

In [13]:
level_mapping = {'low':0, 'medium':1, 'high':2}
df_cat['A2'] = df_cat['A2'].replace(level_mapping)
df_cat.head(5)

A1  A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15
0  b   1  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0
1  a   2  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560
2  a   1  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824
3  b   1  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3
4  b   0  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0

As we can see above the feature 'A2' is converted into a numeric feature with 3 levels due to integer encoding.
<br>
Further more, for rest of the categorical features, we have to use one-hot-encoding technique. For that we have to define **q** dummy variables for a categorical descriptive variable with **q** levels. The exception here is that when a categorical descriptive feature has only two levels, we define a single dummy variable.
<br>
For each two-level categorical variable,  we have to make "drop_first" option "True" . We are using get_dummies() function to perform one-hot-encoding here. 
First, we have to check the categorical variables in the dataframe df_cat.

In [14]:
categorical_cols = df_cat.columns[df_cat.dtypes==object].tolist()
categorical_cols

['A1', 'A4', 'A5', 'A6', 'A7', 'A9', 'A10', 'A12', 'A13', 'A14']

In [15]:
for col in categorical_cols:
    n = len(df_cat[col].unique())
    if (n == 2):
        df_cat[col] = pd.get_dummies(df_cat[col], drop_first=True) ## for categorical features with 2 levels 
# for categorical features with >2 levels        
df_cat_encoded = pd.get_dummies(df_cat)

After encoding, the feature set has the following columns: 

In [16]:
df_cat_encoded.columns

Index(['A1', 'A2', 'A3', 'A8', 'A9', 'A10', 'A11', 'A12', 'A15', 'A4_l',
       ...
       'A14_00640', 'A14_00680', 'A14_00711', 'A14_00720', 'A14_00760',
       'A14_00840', 'A14_00928', 'A14_00980', 'A14_01160', 'A14_02000'],
      dtype='object', length=211)

We can see that, all the columns are encoded into numeric values after one-hot-encoding. The number of features increases to 211 as for some features, multiple levels of unique feature values are present in the data.

## Scaling of Features 

After encoding of the categorical values, we have to scale values of the descriptive features using standard scaling. Standard scaling process includes scaling by removing mean and then dividing by standard deviation of the respective feature column. After the scaling, each descriptive feature has mean of 0  and SD of 1.
<br>
For this purpose we are using "preprocessing" module from the "sklearn" library.

In [17]:
from sklearn import preprocessing
Data_scaler = preprocessing.StandardScaler()
df_cat_encoded_scaled = Data_scaler.fit_transform(df_cat_encoded)

After scaling is done, this process will convert our dataframe into numpy n dimensional array (numpy.ndarray).
<br>
As the output of the scaler is ndarray object, all the column names are lost. so we have to convert this array again into data frame and assign previous data frames column names.

In [18]:
df_cat_encoded_scaled_df = pd.DataFrame(df_cat_encoded_scaled, columns = df_cat_encoded.columns)

In [19]:
df_clean = round(df_cat_encoded_scaled_df,3)

In [20]:
df_clean['target'] = Target

In [21]:
df_clean.shape

(690, 212)

In [22]:
df_clean.describe(include = "all").round(3)

A1       A2       A3       A8       A9      A10      A11      A12  \
count  690.000  690.000  690.000  690.000  690.000  690.000  690.000  690.000   
mean    -0.000   -0.000    0.000    0.000   -0.000    0.000   -0.000    0.000   
std      1.001    1.001    1.001    1.001    1.001    1.001    1.001    1.001   
min     -1.512   -1.215   -0.957   -0.665   -1.048   -0.864   -0.494   -0.919   
25%     -1.512   -1.215   -0.756   -0.616   -1.048   -0.864   -0.494   -0.919   
50%      0.661    0.007   -0.404   -0.366    0.955   -0.864   -0.494   -0.919   
75%      0.661    1.229    0.493    0.120    0.955    1.157    0.123    1.088   
max      0.661    1.229    4.672    7.858    0.955    1.157   13.294    1.088   

           A15     A4_l   ...     A14_00680  A14_00711  A14_00720  A14_00760  \
count  690.000  690.000   ...       690.000    690.000    690.000    690.000   
mean     0.000   -0.000   ...         0.000      0.000     -0.000      0.000   
std      1.001    1.001   ...         1.001      1.001      1.001      1.001   
min     -0.195   -0.054   ...        -0.038     -0.038     -0.054     -0.038   
25%     -0.195   -0.054   ...        -0.038     -0.038     -0.054     -0.038   
50%     -0.194   -0.054   ...        -0.038     -0.038     -0.054     -0.038   
75%     -0.119   -0.054   ...        -0.038     -0.038     -0.054     -0.038   
max     19.012   18.547   ...        26.249     26.249     18.547     26.249   

       A14_00840  A14_00928  A14_00980  A14_01160  A14_02000   target  
count    690.000    690.000    690.000    690.000    690.000  690.000  
mean       0.000      0.000      0.000      0.000      0.000    0.445  
std        1.001      1.001      1.001      1.001      1.001    0.497  
min       -0.038     -0.038     -0.038     -0.038     -0.038    0.000  
25%       -0.038     -0.038     -0.038     -0.038     -0.038    0.000  
50%       -0.038     -0.038     -0.038     -0.038     -0.038    0.000  
75%       -0.038     -0.038     -0.038     -0.038     -0.038    1.000  
max       26.249     26.249     26.249     26.249     26.249    1.000  

[8 rows x 212 columns]

In [23]:
df_clean.head(5)

A1     A2     A3     A8     A9    A10    A11    A12    A15   A4_l  \
0  0.661  0.007 -0.957 -0.291  0.955  1.157 -0.288 -0.919 -0.195 -0.054   
1 -1.512  1.229 -0.060  0.244  0.955  1.157  0.741 -0.919 -0.088 -0.054   
2 -1.512  0.007 -0.856 -0.216  0.955 -0.864 -0.494 -0.919 -0.037 -0.054   
3  0.661  0.007 -0.647  0.457  0.955  1.157  0.535  1.088 -0.195 -0.054   
4  0.661 -1.215  0.174 -0.154  0.955 -0.864 -0.494 -0.919 -0.195 -0.054   

    ...    A14_00680  A14_00711  A14_00720  A14_00760  A14_00840  A14_00928  \
0   ...       -0.038     -0.038     -0.054     -0.038     -0.038     -0.038   
1   ...       -0.038     -0.038     -0.054     -0.038     -0.038     -0.038   
2   ...       -0.038     -0.038     -0.054     -0.038     -0.038     -0.038   
3   ...       -0.038     -0.038     -0.054     -0.038     -0.038     -0.038   
4   ...       -0.038     -0.038     -0.054     -0.038     -0.038     -0.038   

   A14_00980  A14_01160  A14_02000  target  
0     -0.038     -0.038     -0.038       1  
1     -0.038     -0.038     -0.038       1  
2     -0.038     -0.038     -0.038       1  
3     -0.038     -0.038     -0.038       1  
4     -0.038     -0.038     -0.038       1  

[5 rows x 212 columns]

In [24]:
df_clean.to_csv("df_clean.csv", index =False)

***

# Question 2 : Examples on K-Nearest Neighbour Algorithm

**Question 1 :  What value would a 3-nearest neighbor prediction model using
Manhattan distance return for the CPI of Russia?**

We have to find out CPI value for Russia as explained in the problem using Manhattan Distance and with K = 3.
<br>
First we have read the data and load it into the python for further processing. 

In [25]:
import numpy as np
import pandas as pd
import math

df1= pd.read_csv("./Asignment1_Q2.csv")# read the data from local file.
df = df1.copy() #Copy the same data into another dataframe as we are droping target row and column from this dataframe to find distances
df2 = df1.copy()
df.tail(5)

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
12       Canada     80.99         24.79          4.9       1.42          14.2   
13    Australia     82.09         25.40          4.2       1.86          11.5   
14       Sweden     81.43         22.18          2.4       1.27          12.8   
15  New Zealand     80.67         27.81          4.9       1.13          12.3   
16       Russia     67.62         31.68         10.0       3.87          12.9   

       CPI  
12  8.6725  
13  8.8442  
14  9.2985  
15  9.4627  
16       ?

In [26]:
CPI_Col = df.iloc[:,6]     # we save the column 'CPI' as it is a target feature column
CPI_Col = CPI_Col.tolist()
df=df.drop('CPI',axis=1)  # we drop the target feature column.

In [27]:
Country_Col = df.iloc[:,0]     #we save the column 'COUNTRY_ID' into another variable and convert it into List.
Country_Col = Country_Col.tolist()
df=df.drop('COUNTRY_ID',axis=1) # we drop the column 'COUNTRY_ID' from df dataframe as we need only numerical data to find distances.

In [28]:
Pred_row = df.iloc[16,:]    # we save the last row of the data for which we have to predict the CPI into another variable and convert it into a list
Pred_row = Pred_row.tolist()
df=df.drop(16, axis = 0)  # we drop the last row as we need it to find distances.

We can see below that after droping the target features and some irrelevant features, we got the dataset from which we can calculate the Manhattan distance of Russia from every other country.

In [29]:
df.tail(5)

LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS
11     80.24         22.07          3.5       1.31          12.0
12     80.99         24.79          4.9       1.42          14.2
13     82.09         25.40          4.2       1.86          11.5
14     81.43         22.18          2.4       1.27          12.8
15     80.67         27.81          4.9       1.13          12.3

To calculate Manhattan Distances from Russia to every other country we need to convert the values into tuple and then take sum of the absolute distances of every feature values related to other countries. 
<br>
Below is the code to calculate Manhattan distances of every country from Russia.

In [30]:
# This code to calculate Manhattan distance of Russia from every country in data frame df:
Pred_row = tuple(Pred_row)
distance_list = []
for i in range(0,16):
    df_row = df.loc[i].tolist()
    df_row = tuple(df_row)
    distance = sum([abs(a-b)for a, b in zip(Pred_row,df_row)])
    distance_list.append(distance)
    
# As we have similar copy of data in dataframe df2, we can add the list of distances to this copy as a new column 
# We name that column as 'DISTANCES'.
df2=df2.drop(16, axis = 0) # before joining the new column we have to drop the target row from the data frame.
df2['DISTANCES']= distance_list
df2.tail(5)

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
11      Germany     80.24         22.07          3.5       1.31          12.0   
12       Canada     80.99         24.79          4.9       1.42          14.2   
13    Australia     82.09         25.40          4.2       1.86          11.5   
14       Sweden     81.43         22.18          2.4       1.27          12.8   
15  New Zealand     80.67         27.81          4.9       1.13          12.3   

       CPI  DISTANCES  
11  8.0461      32.19  
12  8.6725      29.11  
13  8.8442      29.96  
14  9.2985      33.61  
15  9.4627      25.36

The next step is to find 3 nearest neighbours according to the distances, for that we have to sort our data frame in ascending order and then reset the indices accordingly.

In [31]:
df2_sorted = df2.sort_values('DISTANCES')
df2_sorted = df2_sorted.reset_index(drop = True)
df2_sorted.head(5)

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
0    Argentina     75.77         32.30         13.3       0.76          10.1   
1        U.S.A     78.51         29.85          6.3       4.72          13.7   
2        China     74.87         29.98         13.7       1.95           6.4   
3         U.K.     80.09         28.49          4.4       2.59          13.0   
4  New Zealand     80.67         27.81          4.9       1.13          12.3   

      CPI  DISTANCES  
0  2.9961      17.98  
1  7.1357      18.07  
2  3.6356      21.07  
3  7.7751      22.64  
4  9.4627      25.36

As we can see that, the 3 nearest neighbours to Russia are Argentina, USA and China according to the Manhattan distance calculations. So to predict the 'CPI' value of Russia, we have to take average of 'CPI's of these three countries. below is the code to take average.

In [32]:
df2_sorted['CPI']= pd.to_numeric(df2_sorted['CPI'])
x= 0.0
for i in range (0,3):
    x += df2_sorted.iloc[i,6]
x = x/3
x

4.589133333333334

**The value of CPI value predicted is : 4.589133333333334**
<br>
***

**Question 2 : What value would a weighted k-NN prediction model return for the CPI
of Russia? Use k “ 16 (i.e., the full dataset) and a weighting scheme of
the reciprocal of the squared Manhattan distance between the neighbor
and the query.**

we already have a dataframe (df2_sorted) for which we have all the manhattan distances from Russia to the respective country

In [33]:
df2_sorted.head(5)

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
0    Argentina     75.77         32.30         13.3       0.76          10.1   
1        U.S.A     78.51         29.85          6.3       4.72          13.7   
2        China     74.87         29.98         13.7       1.95           6.4   
3         U.K.     80.09         28.49          4.4       2.59          13.0   
4  New Zealand     80.67         27.81          4.9       1.13          12.3   

      CPI  DISTANCES  
0  2.9961      17.98  
1  7.1357      18.07  
2  3.6356      21.07  
3  7.7751      22.64  
4  9.4627      25.36

As it is mentioned in the problem description that we have to predict 'CPI' with weighted KNN method we have to find weights for all the distances of each country. To find weights, we have to take distances in a list and then calculate the reciprocal of their squares.

In [34]:
Weights = df2_sorted['DISTANCES']
Weights_list = 1/Weights**2
Weights_list

0     0.003093
1     0.003063
2     0.002253
3     0.001951
4     0.001555
5     0.001450
6     0.001355
7     0.001262
8     0.001180
9     0.001158
10    0.001114
11    0.000965
12    0.000885
13    0.000114
14    0.000087
15    0.000076
Name: DISTANCES, dtype: float64

As we have value of K as 16 i.e. all the Countries, we have to take summation of Weights of each country and that of product of CPI and respective weight. 

In [35]:
Sum_Weights_list = round(sum(Weights_list),4) # Summation of weights of each country.
Sum_Weights_list

0.0216

In [36]:
df2_sorted['CPI']= pd.to_numeric(df2_sorted['CPI'])
CPI_list = df2_sorted['CPI']
CPI_list
CPI_WEIGHT_list = [a*b for a, b in zip(CPI_list, Weights_list)]
Sum_CPI_WEIGHT_list = round(sum(CPI_WEIGHT_list),4) # summation of product of weights of each country to their respective CPI value.
Sum_CPI_WEIGHT_list

0.132

As we have value of K as 16, we can predict the 'CPI' value of Russia just by dividing the Sum of (CPI*Weights of each country) by sum of (Weights of each country).

In [37]:
round(Sum_CPI_WEIGHT_list/Sum_Weights_list,2)

6.11

**The value of CPI value predicted using weighted KNN with k =16 is: 6.11**
<br>
***

**Question 3 : The descriptive features in this dataset are of different types. For example,
some are percentages, others are measured in years, and others are
measured in counts per 1,000. We should always consider normalizing
our data, but it is particularly important to do this when the descriptive
features are measured in different units. What value would a 3-
nearest neighbor prediction model using Manhattan distance return for
the CPI of Russia when the descriptive features have been normalized
using range normalization?**

We can amke a new copy of our original data frame in df3. We are using standard value of range for this problem as High = 1 and Low  = 0.
<br>
The first step to solve this problem is making the dataframe ready to ccaluculate Manhattan distances by droping some columns and the target row of Russia.

In [38]:
df3 = df1.copy()
df3_CPI = df3.iloc[:,6]
df3_COUNTRY = df3.iloc[:,0]
df3 = df3.drop(['CPI','COUNTRY_ID'], axis = 1)
df3

LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS
0      59.61         23.21         74.3       4.44           0.4
1      45.00         47.67         73.1       0.09           3.4
2      51.30         38.23         82.6       1.07           4.1
3      70.48         26.58         19.6       1.86           5.3
4      75.77         32.30         13.3       0.76          10.1
5      74.87         29.98         13.7       1.95           6.4
6      73.12         42.93         14.5       1.43           7.2
7      81.30         28.80          3.6       6.77          12.5
8      78.51         29.85          6.3       4.72          13.7
9      80.15         27.23          3.5       0.60          11.5
10     80.09         28.49          4.4       2.59          13.0
11     80.24         22.07          3.5       1.31          12.0
12     80.99         24.79          4.9       1.42          14.2
13     82.09         25.40          4.2       1.86          11.5
14     81.43         22.18          2.4       1.27          12.8
15     80.67         27.81          4.9       1.13          12.3
16     67.62         31.68         10.0       3.87          12.9

The next step is to normalize all the data by using Range Normalization process. Range Normalization includes removing the Minimum value of a feature from a given value and then dividing it by range of that feature column.
We are standardizing the values of High and Low as 1 and 0 so the calue of (High - low)*low becomes 1.

In [39]:
df3_cols = df3.columns.tolist()
df4 = pd.DataFrame(columns = df3_cols)
for i in df3_cols :
    df4[i]= (df3[i]-min(df3[i]))/(max(df3[i])-min(df3[i]))
df4

LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS
0   0.393907      0.044531     0.896509   0.651198      0.000000
1   0.000000      1.000000     0.881546   0.000000      0.217391
2   0.169857      0.631250     1.000000   0.146707      0.268116
3   0.686978      0.176172     0.214464   0.264970      0.355072
4   0.829604      0.399609     0.135910   0.100299      0.702899
5   0.805338      0.308984     0.140898   0.278443      0.434783
6   0.758156      0.814844     0.150873   0.200599      0.492754
7   0.978700      0.262891     0.014963   1.000000      0.876812
8   0.903478      0.303906     0.048628   0.693114      0.963768
9   0.947695      0.201563     0.013716   0.076347      0.804348
10  0.946077      0.250781     0.024938   0.374251      0.913043
11  0.950121      0.000000     0.013716   0.182635      0.840580
12  0.970342      0.106250     0.031172   0.199102      1.000000
13  1.000000      0.130078     0.022444   0.264970      0.804348
14  0.982205      0.004297     0.000000   0.176647      0.898551
15  0.961715      0.224219     0.031172   0.155689      0.862319
16  0.609868      0.375391     0.094763   0.565868      0.905797

The next step is to normalize the target row with the same range normalization. and then we drop the target row after normalizing it from the data frame.

In [40]:
df4_Russia = df4.iloc[16,:].tolist()
df4 = df4.drop(16, axis = 0)
df4_Russia = [round(z,6)for z in df4_Russia]
df4_Russia

[0.609868, 0.375391, 0.094763, 0.565868, 0.905797]

To calculate Manhattan Distances from Russia to every other country we need to convert the values into tuple and then take sum of the absolute distances of every feature values related to other countries. 
<br>
Below is the code to calculate Manhattan distances of every country from Russia.

In [41]:
df4_Russia = tuple(df4_Russia)
distance_list = []
for i in range(0,16):
    df_row = df4.loc[i].tolist()
    df_row = tuple(df_row)
    distance = sum([abs(a-b)for a, b in zip(df4_Russia,df_row)])
    distance_list.append(distance)

df4['CPI']=df3_CPI
df4['COUNTRY_ID']= df3_COUNTRY
df4['DISTANCES']= distance_list
df4

LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS     CPI  \
0   0.393907      0.044531     0.896509   0.651198      0.000000  1.5171   
1   0.000000      1.000000     0.881546   0.000000      0.217391  1.7999   
2   0.169857      0.631250     1.000000   0.146707      0.268116  2.4493   
3   0.686978      0.176172     0.214464   0.264970      0.355072  2.8622   
4   0.829604      0.399609     0.135910   0.100299      0.702899  2.9961   
5   0.805338      0.308984     0.140898   0.278443      0.434783  3.6356   
6   0.758156      0.814844     0.150873   0.200599      0.492754  3.7741   
7   0.978700      0.262891     0.014963   1.000000      0.876812  5.8069   
8   0.903478      0.303906     0.048628   0.693114      0.963768  7.1357   
9   0.947695      0.201563     0.013716   0.076347      0.804348   7.536   
10  0.946077      0.250781     0.024938   0.374251      0.913043  7.7751   
11  0.950121      0.000000     0.013716   0.182635      0.840580  8.0461   
12  0.970342      0.106250     0.031172   0.199102      1.000000  8.6725   
13  1.000000      0.130078     0.022444   0.264970      0.804348  8.8442   
14  0.982205      0.004297     0.000000   0.176647      0.898551  9.2985   
15  0.961715      0.224219     0.031172   0.155689      0.862319  9.4627   

     COUNTRY_ID  DISTANCES  
0   Afghanistan   2.339693  
1         Haiti   3.275534  
2       Nigeria   2.657949  
3         Egypt   1.247652  
4     Argentina   0.953568  
5         China   1.066451  
6        Brazil   1.422163  
7        Israel   1.024251  
8         U.S.A   0.596446  
9       Ireland   1.183672  
10         U.K.   0.729507  
11      Germany   1.245142  
12       Canada   1.154176  
13    Australia   1.110111  
14       Sweden   1.234662  
15  New Zealand   1.020267

After calculating the distance between Russia and every other country, the next step is to sort the dataframe according to the distances and shifting the "COUNTRY_ID" column to the first. 

In [42]:
df4_sorted= df4.sort_values('DISTANCES')
df4_sorted = df4_sorted[['COUNTRY_ID']+[col for col in df4_sorted if col not in ['COUNTRY_ID']]]
df4_sorted= df4_sorted.reset_index(drop = True)
df4_sorted

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
0         U.S.A  0.903478      0.303906     0.048628   0.693114      0.963768   
1          U.K.  0.946077      0.250781     0.024938   0.374251      0.913043   
2     Argentina  0.829604      0.399609     0.135910   0.100299      0.702899   
3   New Zealand  0.961715      0.224219     0.031172   0.155689      0.862319   
4        Israel  0.978700      0.262891     0.014963   1.000000      0.876812   
5         China  0.805338      0.308984     0.140898   0.278443      0.434783   
6     Australia  1.000000      0.130078     0.022444   0.264970      0.804348   
7        Canada  0.970342      0.106250     0.031172   0.199102      1.000000   
8       Ireland  0.947695      0.201563     0.013716   0.076347      0.804348   
9        Sweden  0.982205      0.004297     0.000000   0.176647      0.898551   
10      Germany  0.950121      0.000000     0.013716   0.182635      0.840580   
11        Egypt  0.686978      0.176172     0.214464   0.264970      0.355072   
12       Brazil  0.758156      0.814844     0.150873   0.200599      0.492754   
13  Afghanistan  0.393907      0.044531     0.896509   0.651198      0.000000   
14      Nigeria  0.169857      0.631250     1.000000   0.146707      0.268116   
15        Haiti  0.000000      1.000000     0.881546   0.000000      0.217391   

       CPI  DISTANCES  
0   7.1357   0.596446  
1   7.7751   0.729507  
2   2.9961   0.953568  
3   9.4627   1.020267  
4   5.8069   1.024251  
5   3.6356   1.066451  
6   8.8442   1.110111  
7   8.6725   1.154176  
8    7.536   1.183672  
9   9.2985   1.234662  
10  8.0461   1.245142  
11  2.8622   1.247652  
12  3.7741   1.422163  
13  1.5171   2.339693  
14  2.4493   2.657949  
15  1.7999   3.275534

After sorting the data frame, to predict the the 'CPI' value of the target column Russia, we have to calculate the average of first 3 distances as value if K is 3. 

In [43]:
df4_sorted['CPI']= pd.to_numeric(df4_sorted['CPI'])
x= 0.0
for i in range (0,3):
    x += df4_sorted.iloc[i,6]
x = x/3
x

5.968966666666667

**The value of CPI value predicted is : 5.968966666666667**
<br>
***

**Question 4 : What value would a weighted k-NN prediction model—with k “ 16
(i.e., the full dataset) and using a weighting scheme of the reciprocal of
the squared Euclidean distance between the neighbor and the query—
return for the CPI of Russia when it is applied to the range-normalized
data?**

To solve the above problem,  we already have a dataframe of the range normalized values of each feature column from our previous problem which is 'df4_sorted'. So, we are using the same dataframe to solve this problem.

In [44]:
df4_sorted.head(5)

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
0        U.S.A  0.903478      0.303906     0.048628   0.693114      0.963768   
1         U.K.  0.946077      0.250781     0.024938   0.374251      0.913043   
2    Argentina  0.829604      0.399609     0.135910   0.100299      0.702899   
3  New Zealand  0.961715      0.224219     0.031172   0.155689      0.862319   
4       Israel  0.978700      0.262891     0.014963   1.000000      0.876812   

      CPI  DISTANCES  
0  7.1357   0.596446  
1  7.7751   0.729507  
2  2.9961   0.953568  
3  9.4627   1.020267  
4  5.8069   1.024251

As it is mentioned in the problem description that we have to predict 'CPI' with weighted KNN method we have to find weights for all the distances of each country. To find weights, we have to take distances in a list and then calculate the reciprocal of their squares.

In [45]:
Weights = df4_sorted['DISTANCES']
Weights_list = 1/Weights**2
Weights_list

0     2.810978
1     1.879061
2     1.099756
3     0.960665
4     0.953208
5     0.879262
6     0.811460
7     0.750682
8     0.713735
9     0.656000
10    0.645004
11    0.642411
12    0.494426
13    0.182676
14    0.141549
15    0.093204
Name: DISTANCES, dtype: float64

For this problem, we have value of K as 16 so to predict the CPI value for Russia, we have to take sum of every product of CPI values to their respective weights and divide it by the summation of every weight calculated.

In [46]:
Sum_Weights_list = round(sum(Weights_list),4) # Summation of weights of each country.
Sum_Weights_list

13.7141

In [47]:
df4_sorted['CPI']= pd.to_numeric(df4_sorted['CPI'])
CPI_list = df4_sorted['CPI']
CPI_list
CPI_WEIGHT_list = [a*b for a, b in zip(CPI_list, Weights_list)]
Sum_CPI_WEIGHT_list = round(sum(CPI_WEIGHT_list),4) # summation of product of weights of each country to their respective CPI value.
Sum_CPI_WEIGHT_list

90.6371

Last step is to take division of the total product of CPI values and respective weights to the summation of total weights for every country.

In [48]:
round(Sum_CPI_WEIGHT_list/Sum_Weights_list,2)

6.61

**The value of CPI value predicted using weighted KNN with k =16 and range normalization is: 6.61**
<br>
***

**Question 5 : The actual 2011 CPI for Russia was 2.4488. Which of the predictions made was the
most accurate? Why do you think this was?**

Answer: As we can see the above values, none of the values can predict the accurate CPI for Russia because of two main reasons. 1. As we are calculating the Manhattan distance between every feature column with that of the Russia, we know that Manhattan distance is not the shortest distance between two points but it was measured along the axes at right angles so it gives longer distances between two points. 2. So generally,  Manhattan distance generally works only if the points are arranged in the form of a grid and the problem which we are working on gives more priority to the distance between the points only along with the grids, but not the geometric distance.

***